source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <cstddef>
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a buffer fragment and the buffer of given length the fragment points to with the
+  // guarantee that buffer and fragment are both aligned according to max_align_t.
+  static Envoy::Buffer::BufferFragmentImpl*
+  allocateBufferAndFragment(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance with first |length| bytes in it.
+  // Data will be moved from |buffer| to this mem slice.
+  // Prerequisite: |buffer| has at least |length| bytes of data and not empty.
+  explicit QuicMemSliceImpl(Envoy::Buffer::Instance& buffer, size_t length);
+
+  QuicMemSliceImpl(const QuicMemSliceImpl& other) = delete;
+  // Move constructors. |other| will not hold a reference to the data buffer
+  // after this call completes.
+  QuicMemSliceImpl(QuicMemSliceImpl&& other) noexcept { *this = std::move(other); }
+
+  QuicMemSliceImpl& operator=(const QuicMemSliceImpl& other) = delete;
+  QuicMemSliceImpl& operator=(QuicMemSliceImpl&& other) noexcept {
+    if (this != &other) {
+      single_slice_buffer_.move(other.single_slice_buffer_);
+    }
+    return *this;
+  }
+
+  // Below methods implements interface needed by QuicMemSlice.
+  void Reset() { single_slice_buffer_.drain(length()); }
+
+  // Returns a char pointer to the one and only slice in buffer.
+  const char* data() const;
+
+  size_t length() const { return single_slice_buffer_.length(); }
+  bool empty() const { return length() == 0; }
+
+private:
+  // Used to align both fragment and buffer at max aligned address.
+  struct BufferFragmentBundle {
+    static BufferFragmentBundle* createBundleWithSize(size_t length) {
+      return reinterpret_cast<BufferFragmentBundle*>(
```

Comment:
Can we have this method construct the object, rather than doing it at the call-site?

```
    static BufferFragmentBundle* createBundleWithSize(size_t length) {
      return new (new char[sizeof(BufferFragmentBundle) + length]) BufferFragmentBundle;
    }
```

I'm trying now to unravel the memory management logic, some of which it looks like you already merged into buffer_impl.h, the releasing logic looks interesting.

Reply:
I played around with this to try to clean it up. PTAL.

https://github.com/jmarantz/envoy/commit/0ec37556101c6258bac32779a71f4e292386739a

Reply:
Let me also explain why I made a few changes:
 - wanted to make it look more like vanilla C++ except with variable-length structures.
 - the existing code would crash if someone added a member variable ahead of the Envoy::Buffer::FragmentImpl. In the new code the ability to simply delete the fragment is implicit in the fact that I used inheritance.
 - if someone added a destructor to Envoy::Buffer::FragmentImpl previously it would not have been called.

Reply:
simplified slightly: https://github.com/jmarantz/envoy/commit/55fc55b9b53277f645e2db04dd45160371902256

Reply:
commented on jmarantz@0ec3755.

Reply:
I saw #7063 is generalizing such memory allocation into an interface. I would like to figure out how to apply InlineStorage in #7063 to this PR.
If we make BufferFragmentBundle inherits from InlineStorage, it would completely shut the door of using provided QuicBufferAlloctor in the future. Though I switched from using customized allocator to using raw new and delete, but that's based on the assumption that we can easily to switch back once we have such need.
Actually what BufferFragmentBundle + InlineStorage will do would be implementing OwnedSlice again. If we completely take customized allocator out of consideration, we don't even need to use BufferFragmentImpl at the very beginning, instead, we can use OwnedImpl to allocate its own inline memory.

Reply:
My suggestion is to make this code simple and clean. If there's a demonstrated performance win from a superior general-purpose allocator we can figure out the best way to switch to that, but no need to leave a bunch of dangling wires now.